demo/Diffusion/demo-diffusion.py

```diff
@@ -0,0 +1,473 @@
+            denoising_fp16 (bool):
+                Run the denoising loop (UNet) in fp16 precision.
+                When enabled image quality will be lower but generally results in higher throughput.
+            guidance_scale (float):
+                Guidance scale is enabled by setting as > 1.
+                Higher guidance scale encourages to generate images that are closely linked to the text prompt, usually at the expense of lower image quality.
+            device (str):
+                PyTorch device to run inference. Default: 'cuda'
+            output_dir (str):
+                Output directory for log files and image artifacts
+            hf_token (str):
+                HuggingFace User Access Token to use for downloading Stable Diffusion model checkpoints.
+            verbose (bool):
+                Enable verbose logging.
+            profile (bool):
+                Insert NVTX profiling markers.
+        """
+
+        if image_height % 8 != 0 or image_width % 8 != 0:
+            raise ValueError(f"Image height and width have to be divisible by 8 but specified as: {image_height} and {image_width}.")
+        # Spatial dimensions of latent tensor
+        self.latent_height = image_height // 8
+        self.latent_width = image_width // 8
+
+        # Only supports single image per prompt.
+        self.num_images = 1
+
+        self.denoising_steps = denoising_steps
+        self.denoising_fp16 = denoising_fp16
+        assert guidance_scale > 1.0
+        self.guidance_scale = guidance_scale
+
+        self.output_dir = output_dir
+        self.hf_token = hf_token
+        self.device = device
+        self.verbose = verbose
+        self.profile = profile
+
+        # A scheduler to be used in combination with unet to denoise the encoded image latens.
+        # This demo uses an adaptation of LMSDiscreteScheduler or DPMScheduler:
+        sched_opts = {'num_train_timesteps': 1000, 'beta_start': 0.00085, 'beta_end': 0.012}
+        if scheduler == "DPM":
+            self.scheduler = DPMScheduler(device=self.device, **sched_opts)
+        elif scheduler == "LMSD":
+            self.scheduler = LMSDiscreteScheduler(device=self.device, **sched_opts)
+        else:
+            raise ValueError(f"Scheduler should be either DPM or LMSD")
+
+        self.tokenizer = None
+
+        self.unet_model_key = 'unet_fp16' if denoising_fp16 else 'unet'
+        self.models = {
+            'clip': CLIP(hf_token=hf_token, image_width=image_width, image_height=image_height, device=device, verbose=verbose),
+            self.unet_model_key: UNet(hf_token=hf_token, image_width=image_width, image_height=image_height, fp16=denoising_fp16, device=device, verbose=verbose),
+            'vae': VAE(hf_token=hf_token, image_width=image_width, image_height=image_height, device=device, verbose=verbose)
+        }
+
+        self.engine = {}
+        self.stream = cuda.Stream()
+
+    def __del__(self):
```

Comment:
`__del__` is not guaranteed to be called. Probably better to use a `free()` function or a context manager